### PR TITLE
Deep zoom (#891)

### PR DESCRIPTION
**Pull request by blms**
## In this PR

- Per #891:
  - Implements deep zoom control (OpenSeadragon) on mobile and desktop

## Questions

- Is the desktop rotation tool MVP? And should it be part of this PR or separate?
  - Decided to make it a separate branch so that I can get this merged first, rotation control is complicated!

### CODE REVIEW

File: geniza/corpus/tests/test_corpus_templates.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def test_viewer(self, client, document):
         response = client.get(document.get_absolute_url())
         assertContains(
             response,
-            '<section id="itt-panel">',
+            '<section id="itt-panel" data-controller="iiif">',
```
Comment (rlskoeser) [inline geniza/corpus/tests/test_corpus_templates.py:58]: why is the controller at this level? shouldn't each deep zoom image be bound to its own instance of the controller?

File: sitemedia/js/controllers/iiif_controller.js [hunk 2]
```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
...
+        if (!OSD) {
+            this.addOpenSeaDragon(
+                container,
+                [container.dataset.iiifUrl],
+                isMobile
+            );
+            OSD = container.querySelector(".openseadragon-container");
+        }
+        // OSD styles have to be set directly on the element instead of adding a class, due to
+        // its use of inline styles
```
Comment (rlskoeser) [inline sitemedia/js/controllers/iiif_controller.js:40]: thanks for the comment. this is an OSD limitation? we can't put our styles on the container?

File: sitemedia/js/controllers/iiif_controller.js [hunk 3]
```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
...
+                // If zooming to less than 100%, force it to 100%
+                zoom = minZoom;
+            }
+            // Set zoom slider value to the chosen percentage
+            zoomSlider.value = parseFloat(zoom);
+            this.updateZoomUI(zoom, false, zoomSlider, zoomSliderLabel);
+        });
+        if (isMobile) {
+            viewer.addHandler("canvas-release", () => {
+                // Handle issue on mobile that causes a scroll trap
```
Comment (rlskoeser) [inline sitemedia/js/controllers/iiif_controller.js:161]: can you elaborate on the scroll trap? Is it if the image is showing mid-screen?

what happens with long images? do we / should we limit them to `100vh` on mobile ?

File: sitemedia/js/controllers/iiif_controller.js [hunk 4]
```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
...
+        if (deactivating) {
+            secondColor = "#9E9E9E";
+            slider.classList.remove("active-thumb");
+        } else if (!slider.classList.contains("active-thumb")) {
+            slider.classList.add("active-thumb");
+        }
+        slider.style.background = `linear-gradient(to right, var(--link-primary) 0%, var(--link-primary) ${percent}%, ${secondColor} ${percent}%, ${secondColor} 100%)`;
+    }
+    resetBounds(viewer) {
+        // Reset OSD viewer to the boundaries of the image, position in top left corner
```
Comment (rlskoeser) [inline sitemedia/js/controllers/iiif_controller.js:182]: is `viewer.viewport.goHome(true);` not sufficient?

File: sitemedia/js/controllers/iiif_controller.js [hunk 5]
```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
...
+                    // reset bounds immediately and zoom to 110%
+                    this.resetBounds(viewer);
+                    viewer.viewport.zoomTo(1.1);
+                }
+            });
+        });
+        viewer.addHandler("zoom", (evt) => {
+            // Handle changes in the canvas zoom
+            let { zoom } = evt;
+            if (zoom <= minZoom) {
```
Comment (rlskoeser) [inline sitemedia/js/controllers/iiif_controller.js:151]: how is it possible to zoom to less than `minZoom`?

**Review by rlskoeser — APPROVED** [review]
The code seems more complicated than I thought would be necessary, but it does seem to be working correctly on both desktop and mobile. I'm ok moving forward with this.

File: sitemedia/js/controllers/iiif_controller.js (hunk 3 above)
Comment (blms) [reply sitemedia/js/controllers/iiif_controller.js:161]: For some reason, on `canvas-release` the OSD container forcibly scrolls into view, which causes the mobile toggle to become hidden. I guess this is to make it easier to manipulate the image, but it makes our control inaccessible.

Long images would be a problem, yes! And the height would likely need to leave space for the toggle control, so maybe something like `100vh - height of toggle` would be a good idea.

File: sitemedia/js/controllers/iiif_controller.js (hunk 5 above)
Comment (blms) [reply sitemedia/js/controllers/iiif_controller.js:151]: You would think it wouldn't be, but it is! Using the scroll wheel on my mouse, I get these two `zoom` events in rapid succession:

```
zoomed to 0.8333333333333334
minZoom: 1

zoomed to 1
minZoom: 1
```

Comment (rlskoeser) [reply sitemedia/js/controllers/iiif_controller.js:151]: ah, weird. I would have thought that the minzoom config in the OSD seadragon would prevent that.

File: sitemedia/js/controllers/iiif_controller.js [hunk 6]
```diff
@@ -4,32 +4,194 @@ import { Controller } from "@hotwired/stimulus";
...
+        // Reset OSD viewer to the boundaries of the image, position in top left corner
+        const bounds = viewer.viewport.getBounds();
+        const newBounds = new OpenSeadragon.Rect(
+            0,
+            0,
+            bounds.width,
+            bounds.height
+        );
+        viewer.viewport.fitBounds(newBounds, true);
+        viewer.viewport.setRotation(0, true);
```
[the remark this answers is not in the document]
Comment (blms) [reply sitemedia/js/controllers/iiif_controller.js:191]: I wasn't familiar with this function, but it sets the bounds incorrectly.

<details><summary>goHome</summary>

Screen Shot 2022-07-07 at 11 21 18 AM
</details>

<details><summary>My function</summary>

Screen Shot 2022-07-07 at 11 21 27 AM
</details>

File: sitemedia/js/controllers/iiif_controller.js (hunk 3 above)
Comment (rlskoeser) [reply sitemedia/js/controllers/iiif_controller.js:161]: ah, ok. please elaborate the comment in the code to explain this so we have it documented when we revisit this code.

The height limit you propose sounds good to me. Gissoo may already have some examples of longer images we can test with, or if not we can ask the team.

File: sitemedia/js/controllers/iiif_controller.js (hunk 2 above)
Comment (rlskoeser) [reply sitemedia/js/controllers/iiif_controller.js:40]: wow, neither option is great. let's leave it as you've implemented for now

File: sitemedia/js/controllers/iiif_controller.js (hunk 3 above)
Comment (rlskoeser) [reply sitemedia/js/controllers/iiif_controller.js]: no, haven't run into it yet — I'm good with tracking that on a separate issue so we can move forward

**Pull request merged**